esp-hal/src/i2c/master/mod.rs

```diff
@@ -1017,7 +1029,7 @@ impl Driver<'_> {
                 let clock = clocks.xtal_clock.convert();
             }
         }
-        self.set_frequency(clock, config.frequency, config.timeout);
+        self.set_frequency(clock, config.frequency, config.timeout)?;
```

Comment:
We should probably split most of this function into two: a validation phase, and a hardware configuration phase. Currently it's possible to fail the process halfway through, but the user can keep using the peripheral with some partially applied configuration. Hardware setup should be infallible.

This doesn't really affect the API, though, so I'm fine with handling this later. https://github.com/esp-rs/esp-hal/issues/2839

Reply:
Yeah, let's keep this PR small and cover validation-related stuff in above mentioned issue